Source/Core/VideoCommon/BPMemory.h

```diff
@@ -1979,6 +1967,93 @@ struct BPS_TmemConfig
   u32 texinvalidate;
 };
 
+struct AllTexUnits;
+
+// The addressing of the texture units is a bit non-obvious.
+// This struct abstracts the complexity away.
+union TexUnitAddress
+{
+  enum class Register : unsigned int
+  {
+    SETMODE0 = 0,
+    SETMODE1 = 1,
+    SETIMAGE0 = 2,
+    SETIMAGE1 = 3,
+    SETIMAGE2 = 4,
+    SETIMAGE3 = 5,
+    SETTLUT = 6,
+    UNKNOWN = 7,
+  };
+
+  struct
+  {
+    u32 UnitIdLow : 2;
+    Register Reg : 3;
+    u32 UnitIdHigh : 1;
+    u32 _pad: 26;
+  };
+
+  TexUnitAddress() = default;
+  TexUnitAddress(u32 unit_id, u32 reg)
+      : UnitIdLow(unit_id & 3), Reg(Register(0)), UnitIdHigh(unit_id >> 2), _pad(0)
+  {
+  }
+
+  static TexUnitAddress FromBPAddress(u32 Address)
+  {
+    // Clear upper two bits (which should always be 0x80)
+    return Common::BitCast<TexUnitAddress>(Address & 0x3f);
+  }
+
+  u32 GetUnitID() const { return UnitIdLow | (UnitIdHigh << 2); }
+
+private:
+  friend AllTexUnits;
+
+  size_t GetOffset() const { return Common::BitCast<u32>(*this); }
+};
+static_assert(sizeof(TexUnitAddress) == sizeof(u32));
+
+// A view of the registers of a single TexUnit
+struct TexUnit
+{
+  TexMode0 texMode0;
+  u32 : 32;
+  u64 : 64;
```

Comment:
Note that [`u32 : 32*3` is legal according to the standard, though MSVC doesn't like it](https://stackoverflow.com/a/39415639/3991344) (and other compilers give warnings).  I've [sent feedback about this](https://developercommunity.visualstudio.com/t/C2034-error-is-in-disagreement-with-the/1527280).

Reply:
Interesting. 

Also, MSVC didn't even like:

```
u64 : 32;
u64 : 64;
```

Which it silently rounded up to 128 bits of padding.

Reply:
On further thought, unless MSVC changes their behavior I'd prefer it if this was written with `u32 : 32` three times, to make it a bit more obvious that it's skipping 3 32-bit fields.